### PR TITLE
Add `diff` to Jest's matcher utils

### PR DESCRIPTION
**Pull request by SimenB**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.) (a bunch of tests failed, though)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/jest/pull/7605 (all exports from `jest-matcher-utils` are added to `this.utils` in matchers here: https://github.com/facebook/jest/blob/080998edd46f1c75e8864dae1e6b6cad99265a8e/packages/expect/src/index.js#L237-L252)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

image

### CODE REVIEW

File: types/jest/index.d.ts
Comment (SimenB) [inline types/jest/index.d.ts]: do I need to declare this dependency anywhere?

Comment (jwbay) [reply types/jest/index.d.ts]: I've not introduced a dependency before either, but looking at this https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest-when/index.d.ts I think you need that triple-slash reference

Comment (jwbay) [reply types/jest/index.d.ts]: The main issue with all the tests blowing up is that this file is now treated as a module by the type checker now that it has a top-level `import`, so its types are no longer global. Thus the thousands of "cannot find 'expect'" errors. You should be able to use a local/dynamic import type. I can't remember the exact syntax, or even the right name for it. But something like:
```ts
diff: typeof import('jest-diff');
```

The triple-slash may or may not be needed. I know some types have package.jsons with dependencies specified.

Comment (SimenB) [reply types/jest/index.d.ts]: https://github.com/DefinitelyTyped/DefinitelyTyped#im-writing-a-definition-that-depends-on-another-definition-should-i-use-reference-types--or-an-import

I'll try the `typeof import` 🙂

Comment (SimenB) [reply types/jest/index.d.ts]: oh, adding `dependencies: {"@types/jest-diff"}` throws an error saying package shouldn't depend on stuff from `types`, or something like that, I forget

**Review by jwbay — APPROVED** [review]

**Review by r3nya — APPROVED** [review]

**Comment by domoritz** [comment]
This breaks a bunch of my builds because we now require `@types/jest-diff` as an additional dependency. Was that intended? Wouldn't it be better to at least use a minor version bump?

**Comment by rubensworks** [comment]
I am having the same problem as @domoritz. Pretty sure the `@types/jest-diff` dependency shouldn't be needed. I would assume that `@types/jest-diff` has a dependency on `@types/jest`, but not the other way around.

**Comment by SimenB** [comment]
the intention was for `@types/jest` to depend on `@types/jest-diff`. There is no reason for `@types/jest-diff` to depend on `@types/jest`. I asked how to declare the dependency higher up, happy to submit another PR fixing it. Seems like the auto-dep system breaks with `typeof import`?

**Comment by rubensworks** [comment]
> the intention was for `@types/jest` to depend on `@types/jest-diff`.

Ah, I see.

I think that a `package.json` file with a dependency on `@types/jest-diff` would be needed here, like this: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pikaday/package.json
But perhaps a maintainer (@minestarks) can confirm this?

**Comment by SimenB** [comment]
You get a linting error when adding a dependency on a `@types/` dependency

**Comment by jonasholtkamp** [comment]
This is definitely breaking (just happened to me) and therefore shouldn't be just a patch-level version change. Even if you don't fix the imports it would be good to have this as a minor or even a major change.

**Comment by SimenB** [comment]
How this should ever be considered a breaking change is beyond me (I know it's broken now, but that's a bug due to a missing dependency declaration). This is a purely additive change

**Comment by sirian** [comment]
@SimenB this is a breaking change because now it brokes builds. It's old story - you should add @types/jest-diff to dependencies 
https://stackoverflow.com/questions/45176661/how-do-i-decide-whether-types-goes-into-dependencies-or-devdependencies

image

**Comment by SimenB** [comment]
>this is a breaking change because now it brokes builds.

Answer:

>I know it's broken now, but that's a bug due to a missing dependency declaration

>It's old story - you should add @types/jest-diff to dependencies

Answer:

>You get a linting error when adding a dependency on a `@types/` dependency

---

Someone with more knowledge than me about how DefinitelyTyped works needs to fix this, I think. Might be a good idea to revert in the meantime?

**Comment by sirian** [comment]
@SimenB look how it's done in @types/mongodb (it depends on @types/bson)

**Comment by SimenB** [comment]
Maybe I need a `/// <reference types="jest-diff" />` at the top? I can send a PR for it, not sure if it fixes the issue or not

**Comment by SimenB** [comment]
#33030

**Comment by jonasholtkamp** [comment]
You can try that on your local machine. Just import the package and compile the project. It'll fail if you don't add `@types/jest-diff` to your dependencies.

**Comment by SimenB** [comment]
The point is that it should add a `dependencies` entry i the generated package.json

**Comment by minestarks** [comment]
Sorry everyone - I missed this issue before I merged. Looking at the other PR now

**Comment by sandersn** [comment]
types-publisher maintainer here! @SimenB you are right, the auto-dependency system doesn't understand import types -- it's one of the TODOs that got dropped when I took over the project. I added a bug for it: https://github.com/Microsoft/types-publisher/issues/575.

**Comment by SimenB** [comment]
Thanks @sandersn!

**Pull request merged**
